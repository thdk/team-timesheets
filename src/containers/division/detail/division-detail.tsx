import React, { useCallback } from "react";
import { TextField } from "@rmwc/textfield";

import { Form, FormField } from "../../../components/layout/form";
import { FlexGroup } from "../../../components/layout/flex";
import IconPicker from "../../../components/icon-picker";

type Props = {
    readonly onNameChanged: (name: string) => void;
    readonly onIconChanged: (icon: string) => void;
    readonly icon?: string;
    readonly name?: string;
}

export const DivisionDetail = (props: Props) => {
    const {
        icon,
        name,
        onIconChanged,
        onNameChanged,
    } = props;

    const handleNameChanged = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
        onNameChanged(e.target.value);
    }, [onNameChanged]);

    const handleIconChanged = useCallback((icon: string) => {
        onIconChanged(icon);
    }, [onIconChanged]);

    return <>
        <Form>
            <FlexGroup extraCssClass="row">
                <FormField>
                    <TextField
                        outlined
                        label="Name"
                        value={name}
                        onChange={handleNameChanged}
                    />
                </FormField>
                <FormField first={false}>
                    <IconPicker
                        onChange={handleIconChanged}
                        value={icon}
                        outlined
                    />
                </FormField>
            </FlexGroup>
        </Form>
    </>;
};
DivisionDetail.displayName = "DivisionDetail";