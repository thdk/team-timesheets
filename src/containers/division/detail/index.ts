export { DivisionDetailContainer as DivisionDetail } from "./division-detail-container";
