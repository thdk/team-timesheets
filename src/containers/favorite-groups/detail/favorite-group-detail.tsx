import React, { useState, useCallback } from "react";
import { Form, FormField } from "../../../components/layout/form";
import { FlexGroup } from "../../../components/layout/flex";
import { TextField } from "@rmwc/textfield";
import { IFavoriteRegistrationGroup } from "../../../../common";
import { FavoriteGroupSelect } from "../select";
import { Checkbox } from "@rmwc/checkbox";

export const FavoriteGroupDetailForm = ({
    groups,
    group: {
        name,
        id,
    },
    onNameChanged,
}: {
    groups: (IFavoriteRegistrationGroup & { id: string })[];
    group: Partial<(IFavoriteRegistrationGroup & { id: string })>,
    onNameChanged: (name: string) => void,
}) => {

    const [overwriteExistingGroup, setOverwriteExistingGroup] = useState(false);
    const [newId, setNewId] = useState(id);

    const handleIdChanged = useCallback((id: string) => {
        const existingGroup = groups.find((group) => group.id === id);
        if (existingGroup) {
            onNameChanged(existingGroup.name);
        }
        setNewId(id);
    }, [groups]);

    return (
        <Form style={{ paddingBottom: 0, paddingTop: "1em" }}>
            <FlexGroup className="row">
                <FormField>
                    {groups && overwriteExistingGroup
                        ? (
                            <FavoriteGroupSelect
                                onChange={handleIdChanged}
                                value={newId}
                            />
                        ) : (
                            <TextField
                                autoFocus={true}
                                outlined
                                label="Name"
                                value={name}
                                onChange={(e: React.ChangeEvent<HTMLInputElement>) => onNameChanged(e.target.value)}
                            />

                        )}
                </FormField>
                <FormField
                    first={false}
                >
                    <Checkbox
                        label="Overwrite existing"
                        checked={overwriteExistingGroup}
                        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                            setOverwriteExistingGroup(e.target.checked);
                        }}
                    >

                    </Checkbox>
                </FormField>

            </FlexGroup>
        </Form>
    );
};
