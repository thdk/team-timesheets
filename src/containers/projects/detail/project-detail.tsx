import React, { useCallback } from "react";
import { TextField } from "@rmwc/textfield";

import { Form, FormField } from "../../../components/layout/form";
import { FlexGroup } from "../../../components/layout/flex";
import UserSelect from "../../users/select";
import IconPicker from "../../../components/icon-picker";

type Props = {
    readonly onNameChanged: (name: string) => void;
    readonly onIconChanged: (icon: string) => void;
    readonly onOwnerChanged: (userId: string) => void;
    readonly icon?: string;
    readonly name?: string;
    readonly ownerId?: string;
}

const ProjectDetail = (props: Props) => {

    const {
        icon,
        name,
        ownerId,
        onOwnerChanged,
        onIconChanged,
        onNameChanged,
    } = props;

    const handleNameChanged = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
        onNameChanged(e.target.value);
    }, [onNameChanged]);

    const handleIconChanged = useCallback((icon: string) => {
        onIconChanged(icon);
    }, [onIconChanged]);

    return <>
        <Form>
            <FlexGroup className="row">
                <FormField>
                    <TextField
                        outlined
                        label="Name"
                        value={name}
                        onChange={handleNameChanged}
                    />
                </FormField>
                <FormField first={false}>
                    <IconPicker
                        onChange={handleIconChanged}
                        value={icon}
                        outlined
                    />
                </FormField>
            </FlexGroup>
            <FlexGroup className="row">
                <FormField>
                    <UserSelect
                        label={"Project owner"}
                        value={ownerId}
                        onChange={onOwnerChanged}
                    />
                </FormField>
            </FlexGroup>
        </Form>
    </>;
};

export default ProjectDetail;