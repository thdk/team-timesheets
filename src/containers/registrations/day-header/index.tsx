import React from "react";
import moment from "moment";
import { goToOverview, goToNewRegistration } from "../../../internal";
import classNames from "classnames";
import FavoriteGroupsMenu from "../../favorite-groups/menu";
import { IRootStore } from "../../../stores/root-store";
import { StoreContext } from "../../../contexts/store-context";

export type GroupedRegistrationHeaderProps = {
    readonly groupKey: string;
    readonly totalTime: number;
    readonly headerClick: (e: React.MouseEvent) => void;
    readonly isCollapsed: boolean;
    readonly isCollapsable?: boolean;
    readonly isMonthView: boolean;
    readonly showAddButton?: boolean;
}

const GroupedRegistrationHeader = (props: GroupedRegistrationHeaderProps) => {
    const store = React.useContext(StoreContext);

    const {
        groupKey,
        totalTime,
        headerClick,
        isCollapsable,
        isCollapsed,
        isMonthView,
        showAddButton = true,
    } = props;

    const goToDate = (store: IRootStore, e: React.MouseEvent, date: Date) => {
        e.preventDefault();
        goToOverview(store, {
            year: date.getFullYear(),
            month: date.getMonth() + 1,
            day: date.getDate()
        }, { track: true })
    };

    const createTotalLabel = (store: IRootStore, date: Date, total?: number) => {
        const dateMoment = moment(date);
        const totalJSX = total !== undefined
            ? <div className="grouped-registration-header-total">
                {parseFloat(total.toFixed(2))} hours
                </div>
            : null;


        const displayJSX = !isMonthView
            ? <>
                <div className="grouped-registration-header-label">
                    Total time: {parseFloat(totalTime.toFixed(2))} hours
                </div>
            </>
            : <>
                <a href="#"
                    onClick={(e: React.MouseEvent) => {
                        e.preventDefault();
                        e.stopPropagation();
                        goToDate(store, e, date);
                    }}
                    className="grouped-registration-header-date"
                >
                    {dateMoment.format("MMMM Do")}
                </a>
                {totalJSX}
            </>;

        return showAddButton
            ? <>
                {displayJSX}
                <FavoriteGroupsMenu
                    onSelect={
                        id => {
                            if (!id) {
                                store.timesheets.toggleSelectedRegistrationDay(date.toDateString(), true);
                                goToNewRegistration(store, dateMoment);
                            } else {
                                store.timesheets.addRegistrations(
                                    store.favorites.favoritesByGroup(id)
                                        .map(reg =>
                                            store.timesheets.copyRegistrationToDate(reg.data!, date)
                                        )
                                );
                            }
                        }
                    }
                ></FavoriteGroupsMenu>
            </>
            : <>
                {displayJSX}
            </>;

    };

    const titleJSX = createTotalLabel(store, new Date(groupKey), totalTime);
    const collapseIconJSX = isCollapsable
        ? <i
            className="grouped-registration-header-icon material-icons mdc-icon-button__icon">
            {isCollapsed ? "chevron_right" : "expand_more"}
        </i>
        : null;

    const cssClass = classNames([
        "grouped-registration-header",
        {
            "grouped-registration-header--collapsable": isCollapsable
        }
    ]);

    return <div
        className={cssClass}
        onClick={headerClick}>
        {collapseIconJSX}
        {titleJSX}
    </div>;
};

export default GroupedRegistrationHeader;