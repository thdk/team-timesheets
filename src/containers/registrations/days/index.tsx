import * as React from 'react';
import { observer } from 'mobx-react-lite';
import { Day as Day } from '../day';
import { StoreContext } from '../../../contexts/store-context';

export enum SortOrder {
    Ascending = 1,
    Descending = -1
}

export interface IDaysProps {
    registrationClick: (id: string) => void;
    registrationToggleSelect?: (id: string) => void;
    totalOnTop?: boolean;
    sortOrder?: SortOrder
    activeDate?: number;
    isCollapsable?: boolean;
    isMonthView: boolean;
    showHeaderAddButton?: boolean;
}

export const Days = observer((props: IDaysProps) => {
    const store = React.useContext(StoreContext);

    const {
        sortOrder = SortOrder.Ascending,
        isMonthView,
        showHeaderAddButton = true,
    } = props;

    return <>{(sortOrder > 0
        ? store.timesheets.registrationsGroupedByDay
        : store.timesheets.registrationsGroupedByDayReversed)
        .map((g, i) => {
            const isCollapsed = !store.timesheets.selectedRegistrationDays
                .some(d => d === g.groupKey);

            return <Day
                key={`group-${i}`}
                group={g}
                {...{ ...props, isCollapsed }}
                headerClick={e => {
                    if ((e.target as HTMLElement).closest(".prevent-propagation"))
                        return;

                    store.timesheets.toggleSelectedRegistrationDay(g.groupKey);
                }}
                isMonthView={isMonthView}
                showHeaderAddButton={showHeaderAddButton}
            />
        })}</>;
});