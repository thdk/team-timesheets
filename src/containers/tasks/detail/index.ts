export { TaskDetailContainer as TaskDetail } from "./task-detail-container";
