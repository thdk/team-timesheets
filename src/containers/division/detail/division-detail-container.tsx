import React, { useCallback } from "react";

import { DivisionDetail } from "./division-detail";
import { useDivisionStore } from "../../../contexts/division-context";
import { observer } from "mobx-react-lite";

export const DivisionDetailContainer = observer(() => {
    const divisionStore = useDivisionStore();
    const activeDocument = divisionStore.activeDocument;

    const handleIconChanged = useCallback((icon: string) => {
        if (activeDocument) {
            activeDocument.icon = icon;
        }
    }, [activeDocument]);

    const handleNameChanged = useCallback((name: string) => {
        if (activeDocument) {
            activeDocument.name = name;
        }
    }, [activeDocument]);

    return activeDocument
        ? <DivisionDetail
            onIconChanged={handleIconChanged}
            onNameChanged={handleNameChanged}
            {...activeDocument}
        />
        : null;
});
