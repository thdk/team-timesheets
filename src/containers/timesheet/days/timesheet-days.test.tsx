import React from "react";
import fs from "fs";
import path from "path";


import { Store } from "../../../stores/root-store";
import { render, waitFor, fireEvent } from "@testing-library/react";
import { TimesheetDays, SortOrder } from ".";
import { Timesheet } from "..";
import { goToNewRegistration } from "../../../routes/registrations/detail";
import { useStore } from "../../../contexts/store-context";
import { initializeTestEnvironment, RulesTestEnvironment, } from "@firebase/rules-unit-testing";
import { act } from "react-dom/test-utils";
import { User } from "firebase/auth";

jest.mock("../../google-calendar");

jest.mock("../../../rules");
jest.mock("../../../routes/registrations/detail");

jest.mock("../../../contexts/store-context");

const projectId = "timesheet-test";

let store: Store;

const setupAsync = async () => {
    store = new Store({
        firestore,
    });
    await store.user.usersCollection.addAsync(
        {
            name: "user 1",
            team: "team-1",
            roles: {
                user: true,
            },
            uid: "user-1",
            divisionId: "",
            recentProjects: [],
            tasks: new Map(),
        },
        "user-1",
    );

    store.auth.setUser({ uid: "user-1" } as User);

    return store;
};

let testEnv: RulesTestEnvironment;
let firestore: any;

beforeAll(async () => {
    testEnv = await initializeTestEnvironment({
        projectId,
        firestore: {
            rules: fs.readFileSync(path.resolve(__dirname, "../../../../firestore.rules.test"), "utf8"),
        }
    });

    firestore = testEnv.unauthenticatedContext().firestore();
});

beforeEach(async () => {
    await setupAsync();
    (useStore as jest.Mock<ReturnType<typeof useStore>>).mockReturnValue(store);
});

afterEach(async () => {
    store.dispose();
    await testEnv.clearFirestore();
});

afterAll(() => testEnv.cleanup());

const registrations = [
    {
        date: new Date(2020, 2, 1),
        userId: "user-1",
        project: "project-1",
        task: "task-1",
        time: 4.5,
        isPersisted: true,
        description: "Foobar 1",
    },
    {
        date: new Date(2020, 2, 2),
        userId: "user-1",
        project: "project-1",
        task: "task-1",
        time: 4.5,
        isPersisted: true,
        description: "Foobar 2",
    },
    {
        date: new Date(2020, 2, 24),
        userId: "user-1",
        project: "project-1",
        task: "task-1",
        time: 2,
        isPersisted: true,
        description: "Foobar 3",
    },
    {
        date: new Date(2020, 2, 24),
        userId: "user-1",
        project: "project-2",
        task: "task-1",
        time: 2,
        isPersisted: true,
        description: "Foobar 4",
    },
    {
        date: new Date(2020, 2, 24),
        userId: "user-1",
        project: "project-2",
        task: "task-2",
        time: 6,
        isPersisted: true,
        description: "Foobar 5",
    },
    {
        date: new Date(2020, 3, 1),
        userId: "user-1",
        project: "project-1",
        task: "task-1",
        time: 2.5,
        isPersisted: true,
        description: "Another month"
    },
];

describe("TimesheetDays", () => {
    it("should render without registrations", () => {
        const { asFragment, unmount } = render(<TimesheetDays
            isMonthView={false}
            registrationClick={jest.fn()}
        />);

        expect(asFragment()).toMatchSnapshot();

        unmount();
    });

    it("should display registrations for the specified month", async () => {
        store.view.setViewDate({
            year: 2020,
            month: 3,
        });

        await Promise.all(
            registrations.map(registration => store.timesheets.addDocument(registration))
        );
        await waitFor(() => expect(store.user.divisionUser?.id).toBeDefined());

        const {
            getByText,
            container,
            unmount,
        } = render(<Timesheet />);

        await waitFor(() => expect(getByText("Foobar 5")));

        let items = container.querySelectorAll(".grouped-registration-header-date");
        expect(
            items.length
        ).toBe(3);

        expect(Array.from(items).map(el => (el as HTMLElement).textContent))
            .toEqual([
                "March 24th",
                "March 2nd",
                "March 1st",
            ]);

        store.timesheets.setRegistrationsGroupedByDaySortOrder(SortOrder.Ascending);

        await waitFor(() => expect(Array.from(container.querySelectorAll(".grouped-registration-header-date")).map(el => (el as HTMLElement).textContent))
            .toEqual([
                "March 1st",
                "March 2nd",
                "March 24th",
            ])
        );

        unmount();
    });

    it("should display registrations for the specified day", async () => {
        store.view.setViewDate({
            year: 2020,
            month: 3,
            day: 24,
        });

        store.auth.setUser({ uid: "user-1" } as User);

        await waitFor(() => expect(store.user.divisionUser).toBeDefined());

        await Promise.all(
            registrations.map(registration => store.timesheets.addDocument(registration))
        );

        const {
            getByText,
            container,
            unmount,
        } = render(<Timesheet />);

        await waitFor(() => expect(getByText("Foobar 5")));

        const items = container.querySelectorAll(".grouped-registration-header");
        expect(
            items.length
        ).toBe(1);

        const registrationItems = container.querySelectorAll(".registration-line");
        expect(
            registrationItems.length
        ).toBe(3);

        unmount();
    });
    it("should unfold registration when header is clicked", async () => {
        store.view.setViewDate({
            year: 2020,
            month: 3,
        });

        store.auth.setUser({ uid: "user-1" } as User);

        await waitFor(() => expect(store.user.divisionUser?.id).toBeDefined());

        await Promise.all(
            registrations.map(registration => store.timesheets.addDocument(registration))
        );

        const {
            getByText,
            container,
            unmount,
        } = render(<Timesheet />);

        await waitFor(() => expect(getByText("Foobar 5")));

        const headerEls = container.querySelectorAll(".grouped-registration-header");
        expect(
            headerEls.length
        ).toBe(3);

        act(() => {
            fireEvent.click(headerEls[1]);
        });

        await waitFor(() => expect(getByText("Foobar 2")));

        unmount();
    });

    it("should NOT unfold registration when header add button is clicked", async () => {
        store.view.setViewDate({
            year: 2020,
            month: 3,
        });

        store.auth.setUser({ uid: "user-1" } as User);

        await Promise.all(
            registrations.map(registration => store.timesheets.addDocument(registration))
        );

        const {
            getByText,
            container,
            unmount,
        } = render(<Timesheet />);

        await waitFor(() => expect(getByText("Foobar 5")));

        const headerAddButtonEls = container.querySelectorAll(".grouped-registration-header-add-button");
        expect(
            headerAddButtonEls.length
        ).toBe(3);

        act(() => {
            fireEvent.click(headerAddButtonEls[1]);
        });

        await waitFor(() => expect(goToNewRegistration).toBeCalled());

        unmount();
    });
});
