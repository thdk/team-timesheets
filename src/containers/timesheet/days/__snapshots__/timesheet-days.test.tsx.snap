// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimesheetDays should render without registrations 1`] = `<DocumentFragment />`;
